MAINT: Move user pointers out of axisdata and simplify iternext

seberg opened the pull request:
This is just because I found the way things were iterated confusing and I think this is a lot less confusing and I can't imagine it's slower.

---

The pointers were always stored for each axis (pointing to the start of the whatever the index into the axis was). I found that confusing: One set of pointers is enough.

But some gufuncs relied on external loop giving a second set of pointers (buffers always do so ufuncs are fine).
So, merged the buffer pointers and the external loop into a new field.

I hope (but not sure) that compilers won't complain about loops that are never taken (didn't notice it locally, but...).

For speed testing, I tried to time with something like:
```
a = np.broadcast_to(np.zeros((2, 4))[:, ::2], (100000, 2, 2))

%timeit np.count_nonzero(a)
%timeit np.count_nonzero(a[..., 0])

%timeit np.nonzero(a)
%timeit np.nonzero(a[..., 0])
```
Especially the first (count nonzero).  That code uses the external loop path without buffering (i.e. what changed most). Locally, saw a small slowdown with nonzero, but on a different box and if anything I think a slight speedup for count-nonzero.

So, I doubt this makes a difference, of course an iteration only C test might see a tiny bit more of a difference.


EDIT: Actually, on my old laptop that has (without turbo-boosting) more stable benchmarks, it's better in all cases.  (of course architectures could behave slightly differently)

seberg committed:
MAINT: Move user pointers out of axisdata and simplify iternext

The pointers were always stored for each axis (pointing to the
start of the whatever the index into the axis was).
I found that confusing: One set of pointers is enough.

But some gufuncs relied on external loop giving a second set of
pointers (buffers always do so ufuncs are fine).
So, merged the buffer pointers and the external loop into a new
field.

I hope (but not sure) that compilers won't complain about
loops that are never taken (didn't notice it locally, but...).

For speed testing, I tried to time with something like:
```
a = np.broadcast_to(np.zeros((2, 4))[:, ::2], (100000, 2, 2))

%timeit np.count_nonzero(a)
%timeit np.count_nonzero(a[..., 0])

%timeit np.nonzero(a)
%timeit np.nonzero(a[..., 0])
```
Especially the first (count nonzero).  That code uses the external
loop path without buffering (i.e. what changed most).
Locally, saw a small slowdown with nonzero, but on a different box
and if anything I think a slight speedup for count-nonzero.

So, I doubt this makes a difference, of course an iteration only
C test might see a tiny bit more of a difference.

mattip commented on numpy/_core/src/multiarray/nditer_constr.c:
```diff
@@ -492,6 +491,11 @@ NpyIter_AdvancedNew(int nop, PyArrayObject **op_in, npy_uint32 flags,
         }
     }
+    else if (itflags&NPY_ITFLAG_EXLOOP)  {
+        /* make sure to update the user pointers (copy to buffer does). */
```
Above as well
```suggestion
        /* make sure to update the user pointers (copy to buffer already does). */
```

seberg replied:
adopted as part of the rebase (in both places).

mattip reviewed (approved):
Makes sense to me, after a rebase and tests pass. The number of lines deleted is more than added, which is always nice, and it cleans up some of the pointer copying back-and-forth.

seberg commented:
OK, rebased, changed the comments slightly, and used/kept `dataptrs` name in one or two places since it is clearer.  (Otherwise, rebase is just straight replacement, although I forgot one place first which led to test failures.)

So going to put this in later once tests pass.  If anyone looks over it and finds an improvement, very happy to follow up!

seberg commented:
If anyones compiler shouts because `*_iternext` functions have unreachable code, please let me know.  I didn't notice it locally or in CI, so I was hoping to not complicate code for that.
